### DISCUSSION TITLE
Json serialisation question

### DISCUSSION THREAD

**Discussion by mutagene** [pydjinni/pydjinni#4] [General]
I'm curious how json serialisation would be best implemented, if one were so inclined. As with cross-langugage-cpp djinni, my interest is principally in serialising and deserialising to/from json in c++.  The way I see it

* Record.Deriving.json ?
* A top level target, like cppjson or something like that?
* A cppjson generator inside of cpp?
* Simply another option for the cpp generator, as in cross-language-cpp djinni

Record.Deriving.json seems the most appealing to me, but it doesn't seem very modular - and if you have a deriving for json, why not protobuf or something else? And how would these derivings be added? Still, it seems somehow quite logical to me - especially given the existence of `str` already which I'm assuming serves or will serve a somewhat similar purpose?

Making a new top level target seems quite modular, but nonsensical. Likewise, although somewhat less so, if as a generator for the cpp target.

Lastly - as an option for the cpp generator - not modular at all. Not sure if there's any advantage over deriving.

Anyway, I'm curious what direction you're headed / what would make most sense to you.

**Comment by jothepro** [comment]
Thanks for the interesting question! I had to think about it for a while, here are my first thoughts:

> Record.Deriving.json seems the most appealing to me, but it doesn't seem very modular

Adding record derivings is not modular indeed! This was a deliberate decision to reduce the complexity of the generator plugins.
I agree that adding an additional `json` deriving for records would be the best way to add json serialization, but I guess a configuration parameter like in the cross-language-cpp djinni would still be required to configure the underlying json library.

> especially given the existence of str already which I'm assuming serves or will serve a somewhat similar purpose?

The `str` deriving will add the capabilities to serialize a record for debugging/logging. Support in [Java](https://github.com/pydjinni/pydjinni/blob/main/src/pydjinni/generator/java/java/templates/record.java.jinja2#L44) and [Objective-C](https://github.com/pydjinni/pydjinni/blob/main/src/pydjinni/generator/objc/objc/templates/source/record.m.jinja2#L48) already exists, support for C++ is planned. I want to add a custom `std::print` formatter for this purpose, but the last time I tried I could not get `std::formatter` to work on all compilers.

>  and if you have a deriving for json, why not protobuf or something else?

Can you elaborate further on this? I assume that you refer to [protobuf support in Snapchat/djinni](https://github.com/Snapchat/djinni#use-protobuf-types-in-djinni)?

In my mind your inquiry boils down to the general question if and how (modular) customization of the generated record dataclasses should be possible. Until now I was kind of hesitant to address this issue because introducing this kind of modularity seems like asking for a lot of complexity.

I haven't yet found a satisfying answer or a direction to address this, but I'll try to keep you updated and I am curious about your further ideas and thoughts!

**Reply by mutagene** [reply]
Thank you for taking the time to make that response. Actually, I haven't really looked into protobuf much in the Snapchat fork (is it for interoperating with gRPC services?) - I was just thinking that if one were to support one form of serialization for records/enums, it could be difficult to justify supporting a specific blessed one, such as json, and not another, such as protobuf or yaml or what-have-you. I mean, it's easy to justify the `str` deriving for debugging as you say, and `parcelable` is a nice fit for Android. Although without also generating AIDL bindings, it seems a bit unfinished somehow.

**Reply by mutagene** [reply]
Ahh, one thing I hadn't thought of when I made my first post was that even if `json` were a supported deriving on records, this would not automatically support nice stringification of enums for serialisation/deserialisation. Actually, not sure how you're handling those with your latest str PR - I guess I should check :)

**Reply by jothepro** [reply]
I don't handle it at all yet. Atm the stringification in C++  only supports some basic datatypes and outputs a `{?}`-placeholder for everything else.
But support for djinni enums/flags is a good idea, this would be a nice bonus!

**Reply by mutagene** [reply]
I've taken a bit more of a look at it, and I am at a bit of an impasse. I thought `deriving` was correct, but unless enum/flags also get this deriving property, what can the strategy be for serializing them? I thought to check if an enum was used as a field of a record, and only generate the serialisation code in that case - as I don't think I have ever used an enum/flag outside of a record in my projects - but of course they *could* very easily be used naked. So... if json serialisation is enabled via yaml/flags, I guess serialisers (or, at least, stringification) should always be generated for enums/flags and only generated for records when specified by `deriving` to allow for types which cannot be serialized.. Maybe? 😅

**Reply by jothepro** [reply]
Yeah, I'm not sure either how to best achieve this. In the context of the custom ostream serialization that I am implementing atm, my first thought was to hardcode custom serialization logic for the used enum directly into the record stringification method where it is needed, similar to what I did it for the date type [here](https://github.com/pydjinni/pydjinni/pull/7/files#diff-db0a77afc313bc8d537ef49df92fd46de30ef0cb4a4f1d8fa75eecbe31c3ede4R80). Stringification of the enum would then only be available in the context of the record where it is used. But if the enum would be used in multiple places, the stringification-logic would just be repeated, producing lots of duplicate strings in the binary 😞. I'm not saying this is a good solution, just my first guess.

I agree it would be cleaner though to define the serialization separately (and only if needed) as you are suggesting.

**Reply by paulocoutinhox** [reply]
Hi,

I think that for now, you can replace the templates and with the record data generate the records with JSON support.

If you use nlohmann/json you use:

```cpp
using json = nlohmann::json;

namespace ns {
    void to_json(json& j, const person& p) {
        j = json{{"name", p.name}, {"address", p.address}, {"age", p.age}};
    }

    void from_json(const json& j, person& p) {
        j.at("name").get_to(p.name);
        j.at("address").get_to(p.address);
        j.at("age").get_to(p.age);
    }
} // namespace ns
```
